cacheops/transaction.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+import threading
+from django.db.transaction import get_connection, Atomic
+from funcy import wraps, once
+
+from .utils import monkey_mix
+
+__all__ = ('in_transaction', 'queue_when_in_transaction', 'install_cacheops_transaction_support')
+
+_marker = object()
+
+
+def _find_latest_context(contexts):
+    if not contexts:
+        return contexts
+    latest = contexts[-1]
+    if isinstance(latest, list):
+        return_value = _find_latest_context(latest)
+        if return_value is not None:
+            return return_value
+        return latest
+    return None
+
+
+def _find_second_latest_context(contexts):
+    if not contexts:
+        return contexts
+    latest = contexts[-1]
+    if isinstance(latest, list):
+        return_value = _find_second_latest_context(latest)
+        if return_value is _marker:
+            return latest
+        if return_value is not None:
+            return return_value
+        return _marker
+    return None
+
+
+def _drop_latest_context(contexts):
+    if not contexts:
+        return True
+    latest = contexts[-1]
+    if isinstance(latest, list):
+        return_value = _drop_latest_context(latest)
+        if return_value:
+            contexts.pop()
+            return False
+        return return_value
+    return True
+
+
+def _flatten_and_ignore_none(items):
+    """
+    returns all non list items in a list of lists of of arbitrary depth at a single depth,
+     ignoring non list items that are `None`.
+    """
+    for item in items:
+        if isinstance(item, list):
+            # todo: use yield from in python 3
+            for x in _flatten_and_ignore_none(item):
+                yield x
+        elif item is not None:
+            yield item
```

Comment:
From [here](http://funcy.readthedocs.org/en/stable/python3.html#writing-cross-python-code):
```python
from funcy.py2 import ikeep, iflatten, is_list
```
However, you won't need this at all after `TransactionQueue` refactor.